File: wumpus-hard/w.pl
#!/usr/bin/perl

my $check_after_player = 0;


my $debug = 0;
my $cnt = 0;
my @directions = ();
my $tx = 0;
my $ty = 0;
my $bx = 0;
my $by = 0;
my $sx = 0;
my $sy = 0;
my $done = 0;
my @grid = ();
my $test = shift;
#open(INFILE,"inputs/in$test");
while(<>){
  my $line = $_;
  chomp($line);
  if($line ne "END")
  {
    if($cnt < 10)
    {
      my @line = split(//,$line);
      my $sub = 0;
      foreach my $lineref (@line)
      {
        $grid[$cnt][$sub] = $lineref;
        if($lineref eq "S")
        {
          $sx = $sub;
          $sy = $cnt;
        }
        elsif($lineref eq "B")
        {
          $bx = $sub;
          $by = $cnt;
        }
        elsif($lineref eq "T")
        {
          $tx = $sub;
          $ty = $cnt;
        }
        $sub++;
      }
    }
    else
    {
      push(@directions,$line);
    }
    $cnt++;
  }
  else
  {
    if($debug)
    {
      for(my $i = 0; $i < 10; $i++)
      {
        for(my $j = 0; $j < 10; $j++)
        {
          print $grid[$i][$j];
        }
        print "\n";
      }
      print "\n";
    }

    if($debug)
    {
      #print @directions;
      print "\n\nCURRENT LOCATION: grid[$sy][$sx]\n";
    }
    foreach my $direction (@directions)
    {
      chomp($direction);
      #my $s = <STDIN>;
      #print "\n$direction\n";
      my ($direction1,$direction2) = split(/ /,$direction);

      if($debug and !$done)
      {
        print "You move $direction1...\n";
      }

      if($direction1 eq "N")
      {
        $sy -= 1;
	if($sy < 0)
        {
          $sy = 9;
        }
      }
      elsif($direction1 eq "S")
      {
        $sy += 1;
        if($sy > 9)
        {
          $sy = 0;
        }
      }
      elsif($direction1 eq "E")
      {
        $sx += 1;
        if($sx > 9)
        {
          $sx = 0;
        }
      }
      else
      {
        $sx -= 1;
        if($sx < 0)
        {
          $sx = 9;
        }
      }
      if($debug and !$done)
      {
        print "\n";
	for(my $i = 0; $i < 10; $i++)
        {
          for(my $j = 0; $j < 10; $j++)
          {
            if(($i == $sy)&&($j == $sx))
            {
              print "X";
            }
            elsif(($i == $by)&&($j == $bx))
            {
              print "Y";
            }
            else
            {
              print $grid[$i][$j];
            }
          }
          print "\n";
        }
      }

      if($check_after_player)
      {
        if($debug and !$done)
        {
          print "check for death\n";
        }

        #print "NEW LOCATION: grid[$sy][$sx]\n";
        if(($sx == $bx)&&($sy == $by))
        {
          if(!$done)
          {
            $done = 1;
            print "DEATH\n";
          }
        }
      }
      if($debug and !$done)
      {
	print "Wumpus moves $direction2...\n";
      }

      if($direction2 eq "N")
      {
        $by -= 1;
	if($by < 0)
        {
          $by = 9;
        }
      }
      elsif($direction2 eq "S")
      {
        $by += 1;
        if($by > 9)
        {
          $by = 0;
        }
      }
      elsif($direction2 eq "E")
      {
        $bx += 1;
        if($bx > 9)
        {
          $bx = 0;
        }
      }
      else
      {
        $bx -= 1;
        if($bx < 0)
        {
          $bx = 9;
        }
      }
      #print "NEW LOCATION: grid[$sy][$sx]\n";

      if($debug and !$done)
      {
        print "\n";
	for(my $i = 0; $i < 10; $i++)
        {
          for(my $j = 0; $j < 10; $j++)
          {
            if(($i == $sy)&&($j == $sx))
            {
              print "X";
            }
            elsif(($i == $by)&&($j == $bx))
            {
              print "Y";
            }
            else
            {
              print $grid[$i][$j];
            }
          }
          print "\n";
        }
      }

      if($debug and !$done)
      {
        print "check for death\n";
      }

      if(($sx == $bx)&&($sy == $by))
      {
        if(!$done)
        {
          $done = 1;
          print "DEATH\n";
        }
      }

      if($debug and !$done)
      {
        print "check for treasure\n";
      }

      if(($sx == $tx)&&($sy == $ty))
      {
        if(!$done)
        {
          $done = 1;
          print "TREASURE\n";
	}
      }
                
    }
    unless((($sx == $bx)&&($sy == $by))||(($sx == $tx)&&($sy == $ty)))
    {
      if(!$done)
      {
        $done = 1;
        print "NOTHING\n";
      }
    }

    undef @grid;
    my @grid = ();
    undef @directions;
    my @directions = ();

    $cnt = 0;
    $tx = 0;
    $ty = 0;
    $bx = 0;
    $by = 0;
    $sx = 0;
    $sy = 0;
    $done = 0;
    if($debug)
    {
      print "END\n";
    }
  }
}	

File: wumpus-hard/w.py
#!/usr/bin/python2.7
#
# problem solution by understudy from quantum koala coalition
# dance@bochs.info

import sys

class Point:
    def __init__(self, x, y):
        self.x = x
        self.y = y

    def move(self, d):
        if d == 'E':
            self.x += 1
        elif d == 'W':
            self.x -= 1
        elif d == 'N':
            self.y -= 1
        elif d == 'S':
            self.y += 1

        self.x = self.x % 10
        self.y = self.y % 10

    def __eq__(self, other):
        return self.x == other.x and self.y == other.y

    def __repr__(self):
        return '(%d, %d)' % (self.x, self.y)

class Grid:
    def __init__(self, grid):
        self.grid = grid
        self.player = self.scan('S')
        self.treasure = self.scan('T')
        self.wumpus = self.scan('B')
        self.status = None

    def scan(self, char):
        grid = self.grid
        for x in xrange(10):
            for y in xrange(10):
                if grid[y][x] == char:
                    return Point(x, y)

    def check(self):
        if self.player == self.wumpus:
            return 'DEATH'
        elif self.player == self.treasure:
            return 'TREASURE'

    def move(self, p, w):
        if self.status:
            return

        self.player.move(p)
        self.status = self.check()
        if self.status:
            return self.status

        self.wumpus.move(w)
        self.status = self.check()
        if self.status:
            return self.status

    def end(self):
        if not self.status:
            return 'NOTHING'

line = True
status = None
game = None
while line:
    if not game:
        grid = [[] for i in xrange(10)]
        for i in xrange(10):
            line = sys.stdin.readline().strip()
            if not line:
                sys.exit(0)
            grid[i].extend(line)
        game = Grid(grid)

    line = sys.stdin.readline().strip()
    if line == 'END':
        status = game.end()
        if status:
            print status
        game = None
    else:
        p, w = line.split(' ', 1)
        status = game.move(p, w)
        if status:
            print status
